File: collector/modules/players.py
"""players — owns `players`: which of the largest anchor investors are in the books of issues on the board today.

Source: InvestorGain report 551 (every anchor investor: IPOs anchored, total invested, average ticket — one call) and
report 561 (one investor's IPOs, newest first, each with our own listing id — one call an investor). The free feed gives
only an investor's FIVE most recent IPOs (`fldpage_size: 5`; page 2 repeats page 1 — a members' limit, left alone), which
is exactly enough for one question: is this investor in a book that is open or about to list? It is NOT enough for a track
record, so none is computed and the page shows none.

So: take the TRACK largest investors by money and by count, read each one's latest five, and turn the lists round into
`books{igId: [investor, ...]}` for listings on the board. The anchor list is published the evening before an issue opens —
this is the one signal that exists before the first bid. Coverage is stated, never implied: a book shows "n of the N
largest anchors"; a small SME book taken by local funds will rightly show none.

Runs in full runs only (about 55 calls, 20 s). A 403 stops the module at once; a failed investor is skipped and counted.
Rows kept from `prev` for listings still on the board when their investor could not be read today.

`frozen{igId}` is the accrual for a track record that CAN be built: the moment an issue lists, its line-up as last read
(how many of the tracked anchors were in, and who) is frozen with the listing date and never touched again. `covered[]`
names every unlisted issue looked up in a run, so "no book" freezes as a genuine zero and not as "never looked".
`evidence` joins frozen with the listing outcome, by segment, with n and an interval — thin at first, and said so.
"""
from __future__ import annotations

import datetime as dt
import logging
import re
from zoneinfo import ZoneInfo

from ..errors import SourceBlocked, SourceChanged, SourceError
from ..http import Session
from ..result import Result
from ..sources import investorgain as ig

log = logging.getLogger("collector.players")
IST = ZoneInfo("Asia/Kolkata")
BOARDS = ("mainboard", "sme")
TRACK_BY_MONEY, TRACK_BY_COUNT = 40, 30
LIST_URL = "https://webnodejs.investorgain.com/cloud/v2/report/data-read/551/1/{m}/{y}/{fy}/0/all"
ONE_URL = "https://webnodejs.investorgain.com/cloud/v2/report/data-read/561/1/{m}/{y}/{fy}/0/{id}"


def _num(v) -> float | None:
    s = re.sub(r"[^\d.]", "", str(v or ""))
    return float(s) if s and s != "." else None


def parse_investors(data) -> list[dict]:
    rows = data.get("reportTableData") if isinstance(data, dict) else None
    if not isinstance(rows, list) or not rows or "AnchorName" not in rows[0] or "~id" not in rows[0]:
        raise SourceChanged(ig.SOURCE, "report 551: no anchor-investor rows", LIST_URL)
    out = [{"id": str(r["~id"]), "name": ig.strip_tags(r.get("AnchorName")), "ipos": int(_num(r.get("~totalcount")) or 0),
            "investedCr": _num(r.get("Total Invested")), "ticketCr": _num(r.get("Avg Ticket Size"))} for r in rows if r.get("~id") and r.get("AnchorName")]
    if not out:
        raise SourceChanged(ig.SOURCE, "report 551: rows without id or name", LIST_URL)
    return out


def parse_ipos(data) -> list[str]:
    """The listing ids in one investor's latest IPOs. An investor with no rows is a changed source: 551 listed them for having some."""
    rows = data.get("reportTableData") if isinstance(data, dict) else None
    if not isinstance(rows, list) or not rows or "~id" not in rows[0]:
        raise SourceChanged(ig.SOURCE, "report 561: no IPO rows for the investor", ONE_URL)
    return [str(r["~id"]) for r in rows if r.get("~id")]


def tracked(investors: list[dict]) -> list[dict]:
    by_money = sorted(investors, key=lambda r: -(r["investedCr"] or 0))[:TRACK_BY_MONEY]
    by_count = sorted(investors, key=lambda r: -r["ipos"])[:TRACK_BY_COUNT]
    seen, out = set(), []
    for r in by_money + by_count:
        if r["id"] not in seen:
            seen.add(r["id"])
            out.append(r)
    return out


def freeze(prev_players: dict, doc: dict, today: dt.date) -> dict:
    """prev's `frozen` plus every issue prev had looked up (`covered`, or a book) that has now listed or left the board:
    {igId: {name, sme, listedOn, large, of, names[]}}. A frozen entry is never rewritten."""
    out = {k: v for k, v in (prev_players.get("frozen") or {}).items() if isinstance(v, dict)}
    books, names = prev_players.get("books") or {}, prev_players.get("names") or {}
    candidates = set(prev_players.get("covered") or []) | set(books)
    if not candidates:
        return out
    rows = {str(r["igId"]): (r, k == "sme") for k in BOARDS for r in (doc.get(k) or []) if isinstance(r, dict) and r.get("igId")}
    perf = {str(r["igId"]): r for r in (doc.get("listedPerf") or []) if isinstance(r, dict) and r.get("igId")}
    for i in candidates:
        if i in out:
            continue
        row, sme = rows.get(i, (None, None))
        if row is not None and row.get("status") != "Listed":
            continue                                        # still on the way: nothing to freeze yet
        p = perf.get(i)
        if row is None and p is None:
            continue                                        # withdrawn or unknown: no outcome will ever join it
        listed_on = (row or {}).get("listing") or (p or {}).get("date")
        if not listed_on:
            continue
        book = books.get(i) or []
        out[i] = {"name": (row or {}).get("name") or (p or {}).get("name") or names.get(i), "sme": bool(sme if row else p.get("sme")),
                  "listedOn": str(listed_on)[:10], "large": len(book), "of": prev_players.get("tracked"),
                  "names": [b.get("name") for b in book if isinstance(b, dict)], "frozenOn": today.isoformat()}
    return out


def run(session: Session, prev: dict, res: Result, today: dt.date | None = None) -> Result:
    t = today or dt.datetime.now(IST).date()
    doc = res.doc or prev
    live = {str(r["igId"]): r["name"] for k in BOARDS for r in (doc.get(k) or [])
            if isinstance(r, dict) and r.get("igId") and r.get("name") and r.get("status") != "Listed"}
    prev_players = prev.get("players") or {}
    frozen = freeze(prev_players, doc, t)
    if not live:
        res.notes.append("no unlisted issue on the board; nothing to look up")
        if frozen != (prev_players.get("frozen") or {}):
            res.replace["players"] = {**prev_players, "frozen": frozen, "covered": []}
        return res.won("none", t.isoformat())
    fy = ig.fiscal_year(t)
    try:
        investors = parse_investors(session.get_json(LIST_URL.format(m=t.month, y=t.year, fy=fy), source=ig.SOURCE, headers=ig.HEADERS))
    except SourceError as e:
        return res.fail(e)
    watch, books, failed, last = tracked(investors), {}, 0, None
    for inv in watch:
        try:
            ids = parse_ipos(session.get_json(ONE_URL.format(m=t.month, y=t.year, fy=fy, id=inv["id"]), source=ig.SOURCE, headers=ig.HEADERS))
        except SourceBlocked as e:                       # a 403 ends the conversation for this run
            return res.fail(e)
        except SourceError as e:
            failed, last = failed + 1, e
            continue
        for i in ids:
            if i in live:
                books.setdefault(i, []).append({k: inv[k] for k in ("id", "name", "ipos", "investedCr", "ticketCr")})
    if failed == len(watch):
        return res.fail(last or SourceChanged(ig.SOURCE, "no investor could be read"))
    old = (prev.get("players") or {}).get("books") or {}
    for i, rows in old.items():                          # a book read yesterday stands when today's pass missed its investors
        if i in live and i not in books and failed:
            books[i] = rows
    for rows in books.values():
        rows.sort(key=lambda r: -(r["investedCr"] or 0))
    res.replace["players"] = {"asOf": t.isoformat(), "tracked": len(watch), "of": len(investors), "latestPerInvestor": 5,
                              "books": books, "names": {i: live[i] for i in books}, "frozen": frozen,
                              "covered": sorted(live) if not failed else sorted(set(books) | {i for i in live if i in old}),
                              "league": [{k: r[k] for k in ("id", "name", "ipos", "investedCr", "ticketCr")} for r in watch[:25]]}
    res.notes.append(f"{len(watch)} of {len(investors)} anchor investors read ({failed} failed); names found for {len(books)} of {len(live)} unlisted issues")
    return res.won("investorgain", t.isoformat())

File: site/src/players.js
/* ================= ANCHORS ON THE BOARD — one line per current issue =================
   Included into app.js (`@include players.js`); shares $, esc, cr, fmtD, days, anchorFor, SEG, EDG, bandOf, edgeLbl, evShort,
   follows. `players` is written by collector/modules/players.py: the N largest anchors' latest five IPOs, turned round
   into books for unlisted issues, and `frozen{igId}` — each issue's line-up the day it listed. `evidence.segments[x].anchors`
   bands those frozen line-ups against the listing outcome (n + Wilson; accruing since Sep 2026, thin and said so).
   No per-investor track record: the feed does not give enough history for one, so none is shown. */
function anchorEvidence(b, large) {
  const seg = SEG(!!b.sme), A = seg && seg.anchors; if (!A || !A.n || large == null) return "";     // nothing frozen yet: the note says so once
  const ed = EDG("anchors"), i = bandOf(large, ed); if (i < 0) return "";
  const s = A.rows[i], lbl = edgeLbl(ed, i, "");
  return `<span class="dt" title="${b.sme ? "SME" : "Mainboard"} listings since ${fmtD(A.since)} whose line-up was frozen the day they listed: ${A.n} on file">past ${b.sme ? "SME" : "mainboard"} books with ${esc(lbl)} large anchors: ${evShort(s)}</span>`;
}
function playerCards(curIss) {
  const P = DATA.players || {}, B = P.books || {}, covered = new Set(P.covered || []);
  const named = new Set(), rows = curIss.slice().sort((x, y) => (x.open || "").localeCompare(y.open || ""));
  const later = [];
  const html = rows.map(b => {
    // `covered` names the issues looked up (from 22 Sep 2026); an older document looked up every unlisted issue
    const id = b.igId != null ? String(b.igId) : null, L = id && B[id] ? B[id] : null, looked = !!L || (id && (P.covered ? covered.has(id) : true));
    if (L) named.add(b.name);
    const A1 = anchorFor(b.name), size = A1 && A1.amountCr ? `${cr(A1.amountCr)}${A1.issueSizeCr ? ` · ${Math.round(A1.amountCr / A1.issueSizeCr * 100)}% of the issue` : ""}` : "";
    const lock = A1 && A1.lockIn30 ? `lock-in ends ${fmtD(A1.lockIn30)} / ${fmtD(A1.lockIn90)}` : "";
    const mine = L ? L.filter(x => follows(x.name)) : [];
    // an Upcoming issue's anchor book is allotted the day before it opens: until then there is no book to be in
    const notYet = !L && b.status === "Upcoming" && !(A1 && A1.amountCr) && !(A1 && A1.date && days(A1.date) < 0);
    const large = L ? L.length : looked && !notYet ? 0 : null;
    if (notYet) { later.push(`<span class="chip-i" data-row data-name="${esc(b.name)}">${esc(b.name)}${A1 && A1.date ? `<small class="dim"> bid ${days(A1.date) === 0 ? "today" : fmtD(A1.date)}</small>` : ""}</span>`); return ""; }
    const who = L ? `<details class="who"><summary>${L.length} of the ${P.tracked} largest anchors are in · who</summary><div class="chips" style="margin-top:6px">${L.map(x => `<span class="${follows(x.name) ? "ok" : ""}">${esc(x.name)}<small class="dim"> ${x.ipos} IPOs</small><button class="fbtn" data-follow="${esc(x.name)}" title="${follows(x.name) ? "Following" : "Follow this name"}">${follows(x.name) ? "✓" : "+"}</button></span>`).join("")}</div></details>`
      : notYet ? `<span class="dt">anchor book not published yet${A1 && A1.date ? ` · bid ${days(A1.date) === 0 ? "today" : fmtD(A1.date)}` : ""}</span>`
      : looked ? `<span class="dt">none of the ${P.tracked} largest anchors in this book</span>` : `<span class="dt">line-up not read yet</span>`;
    return `<div class="arow" data-row data-name="${esc(b.name)}"><div class="an"><b>${esc(b.name)}</b> <span class="pill ${b.status.toLowerCase()}">${esc(b.status)}</span>${b.sme ? ` <span class="pill plan" style="height:17px;font-size:10.5px">SME</span>` : ""}<div class="dt">${[size, lock].filter(Boolean).join(" · ") || (notYet ? "" : "no anchor book on file")}</div></div>
      <div class="aw">${who}${mine.length ? `<div class="dt" style="color:var(--up)">your names: ${mine.map(x => esc(x.name)).join(" · ")}</div>` : ""}${anchorEvidence(b, large)}</div></div>`; }).join("")
    + (later.length ? `<div class="arow" style="grid-template-columns:1fr"><div class="dt" style="margin-bottom:4px">Anchor books still to be published · ${later.length} issue${later.length > 1 ? "s" : ""}</div><div class="wl">${later.join("")}</div></div>` : "");
  return { html, named };
}
const playersNote = () => { const P = DATA.players || {}, A = (SEG(false) || {}).anchors; return P.tracked ? `Names come from the ${P.tracked} largest of ${(P.of || 0).toLocaleString("en-IN")} anchor investors on file (their latest ${P.latestPerInvestor} IPOs each; ${P.asOf ? "read " + fmtD(P.asOf) : ""}). A book with none of them is a book the big funds skipped. Each line-up is frozen the day the issue lists${A && A.n ? `; ${A.n} mainboard line-ups on file since ${fmtD(A.since)}` : "; the record starts now, so the past-books line reads thin for a while"}.` : "Anchor names are not on file yet."; };
